src/mlpack/core/optimizers/nestrov/nestrov_update.hpp

```diff
@@ -0,0 +1,112 @@
+
+#include <mlpack/prereqs.hpp>
+#include "vanilla_momentum.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Construct the NAG optimizer with the given function and parameters.  The
+ * defaults here are not necessarily good for the given problem, so it is
+ * suggested that the values used be tailored to the task at hand.  The
+ * maximum number of iterations refers to the maximum number of points that
+ * are processed (i.e., one iteration equals one point; one iteration does not
+ * equal one pass over the dataset). Typically the momentum paramter is often
+ * initialized with small value like 0.5 and later raised.
+
+ * @misc{1212.0901,
+ * Author = {Yoshua Bengio and Nicolas Boulanger-Lewandowski and Razvan Pascanu},
+ * Title = {Advances in Optimizing Recurrent Networks},
+ * Year = {2012},
+ * Eprint = {arXiv:1212.0901},
+ * } 
+ *
+ * @code
+ * @book{Goodfellow-et-al-2016,
+ *  title={Deep Learning},
+ *  author={Ian Goodfellow and Yoshua Bengio and Aaron Courville},
+ *  publisher={MIT Press},
+ *  note={\url{http://www.deeplearningbook.org}},
+ *  year={2016}
+ * }
+ */
+class NestrovUpdate
+{
+ public:
+  /**
+   * Construct the momentum update policy with given momentum decay parameter.
+   *
+   * @param momentum The momentum decay hyperparameter
+   */
+  NestrovUpdate(const double momentum = 0.5): momentum(momentum)
+  {};
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start of
+   * the iteration update process.  In the momentum update policy the velocity
+   * matrix is initialized to the zeros matrix with the same size as the
+   * gradient matrix (see mlpack::optimization::SGD::Optimizer )
+   *
+   * @param n_rows number of rows in the gradient matrix.
+   * @param n_cols number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows,
+                  const size_t cols)
+  {
+    //Initialize am empty velocity matrix.
+    velocity = arma::zeros<arma::mat>(rows, cols);
+  }
+
+  /**
+   * Update step for SGD.  The momentum term makes the convergence faster on the
+   * way as momentum term increases for dimensions pointing in the same and
+   * reduces updates for dimensions whose gradients change directions.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  template<typename... T>
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient,
+              T... args)
+  {
+    double momentumLookAhead;
+    velocity = momentum * velocity - stepSize * gradient;
+    UpdateMomentum(momentum, momentumLookAhead);
+    iterate += momentumLookAhead * momentum * velocity 
+      - (1 + momentumLookAhead) * stepSize * gradient;
+  }
+
+  template<typename DecayPolicy = VanillaMomentum>
+  void UpdateMomentum(const double& momentum, double& momentumLookAhead)
```

Comment:
Okay so basically if somebody wants to have a new policy he would have to go ahead build another policy for the momentum update. Something decay_momentum.hpp. I see that this would not be very useful. Should I provide some more decay policies. Or rather have the UpdateMomentum accept and function that can be implemented by the user that takes momentum and returns the updated momentum. Let me know what you think?